pkg/services/ngalert/notifier/external_alertmanager.go

```diff
@@ -121,30 +123,62 @@ func (am *externalAlertmanager) ListSilences(ctx context.Context, filter []strin
 	return res.Payload, nil
 }
 
-func (am *externalAlertmanager) GetStatus() apimodels.GettableStatus {
-	return apimodels.GettableStatus{}
+func (am *externalAlertmanager) GetAlerts(ctx context.Context, active, silenced, inhibited bool, filter []string, receiver string) (apimodels.GettableAlerts, error) {
+	params := amalert.NewGetAlertsParamsWithContext(ctx).
+		WithActive(&active).
+		WithSilenced(&silenced).
+		WithInhibited(&inhibited).
+		WithFilter(filter).
+		WithReceiver(&receiver)
+
+	res, err := am.amClient.Alert.GetAlerts(params)
+	if err != nil {
+		return apimodels.GettableAlerts{}, err
+	}
+
+	return res.Payload, nil
 }
 
-func (am *externalAlertmanager) GetAlerts(active, silenced, inhibited bool, filter []string, receiver string) (apimodels.GettableAlerts, error) {
-	return apimodels.GettableAlerts{}, nil
+func (am *externalAlertmanager) GetAlertGroups(ctx context.Context, active, silenced, inhibited bool, filter []string, receiver string) (apimodels.AlertGroups, error) {
+	params := amalertgroup.NewGetAlertGroupsParamsWithContext(ctx).
+		WithActive(&active).
+		WithSilenced(&silenced).
+		WithInhibited(&inhibited).
+		WithFilter(filter).
+		WithReceiver(&receiver)
+
+	res, err := am.amClient.Alertgroup.GetAlertGroups(params)
+	if err != nil {
+		return apimodels.AlertGroups{}, err
+	}
+
+	return res.Payload, nil
 }
 
-func (am *externalAlertmanager) GetAlertGroups(active, silenced, inhibited bool, filter []string, receiver string) (apimodels.AlertGroups, error) {
-	return apimodels.AlertGroups{}, nil
+func (am *externalAlertmanager) PutAlerts(ctx context.Context, postableAlerts apimodels.PostableAlerts) error {
+	alerts := make(alertingNotify.PostableAlerts, 0, len(postableAlerts.PostableAlerts))
+	for _, pa := range postableAlerts.PostableAlerts {
+		alerts = append(alerts, &alertingNotify.PostableAlert{
+			Annotations: pa.Annotations,
+			EndsAt:      pa.EndsAt,
+			StartsAt:    pa.StartsAt,
+			Alert:       pa.Alert,
+		})
+	}
+
+	params := amalert.NewPostAlertsParamsWithContext(ctx).WithAlerts(alerts)
+	_, err := am.amClient.Alert.PostAlerts(params)
+	return err
```

Comment:
This is not the implementation that we want here, I would prefer if we leave it out of this PR and just let this PR handle reads.

I'll leave you with the details so that you can file an issue and add it as a follow-up.

If you look closely at how prometheus handles sending of alerts to the alertmanager, you'll realise it's a combination of two things:

- [The `SendAlerts` function](https://github.com/prometheus/prometheus/blob/main/rules/manager.go#L1252-L1276)
- [The `Notifier` struct](https://github.com/prometheus/prometheus/blob/main/notifier/notifier.go#L349-L379)

This is important because the notifier uses an internal queue to make sure alerts are delivered in batches when possible.

The good news is that [we already have this implementation in Grafana](https://github.com/grafana/grafana/blob/main/pkg/services/ngalert/sender/notifier.go), the bad news is that we'll need to adapt it so that it works with or in the `externalAlertmanager`. Please note that the way that "remote URLs" are configured in the notifier is not straightforward and require some level of finesse to implement as they rely on service discovery, to top that off, the notifier is used in other parts of Grafana (hence the public API).

Reply:
Issue created https://github.com/grafana/grafana/issues/76692